pkg/repository/repository.go

```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022-2023 Dell Inc, or its subsidiaries.
+
+// Package repository is the database abstraction implementing repository design pattern
+package repository
+
+// Database abstraction
+type Database interface {
+	Set(key string, value string) (string, error)
+	Get(key string) (string, error)
+	Delete(key string) (string, error)
+}
+
+// Factory pattern to create new Database
+func Factory(databaseImplementation string) (Database, error) {
```

Comment:
You probably want to define a database enum instead of using a string

Reply:
For databases that are backed by a provided service, i.e. Redis (or anything not in memory or a persistent volume), you will need to pass the url, host, port, user, pass, etc... I think it would be worth creating a `config` struct in the `repository` package for these values and passing that to the factory too. All keys would be optional and you would rely on code documentation to state which keys are required for each implementation.

You could include the `databaseImplementation` value as part of this `config` struct too.

Depending on the flexibility required, you could create a `config` struct based on a free-form `map[string]string`. This wouldn't force you to pre-define the keys required for a `Database` instance, but instead you would rely on database implementations clearly documenting the required key-value config entries required. An example of this is here: https://matthewbrown.io/2016/01/23/factory-pattern-in-golang